core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+                    .collect(toImmutableList());
+            node.getSourceFragmentIds()
+                    .forEach(sourceFragmentId -> fragmentTopology.addEdge(sourceFragmentId, currentFragmentId));
+            return new Fragments(
+                    sources.stream()
+                            .flatMap(source -> source.getAllFragments().stream())
+                            .collect(toImmutableSet()),
+                    sources.stream()
+                            .flatMap(source -> source.getLazyFragments().stream())
+                            .collect(toImmutableSet()),
+                    // initially current fragment is considered to be lazy unless there exist
+                    // an operator that can fully consume input data without producing any output
+                    // (e.g. final aggregation)
+                    true);
+        }
+
+        @Override
+        public Fragments visitExchange(ExchangeNode node, PlanFragmentId currentFragmentId)
+        {
+            checkArgument(node.getScope() == LOCAL, "Only local exchanges are supported in the phased execution scheduler");
+            return visitPlan(node, currentFragmentId);
+        }
+
+        @Override
+        protected Fragments visitPlan(PlanNode node, PlanFragmentId currentFragmentId)
+        {
+            List<PlanNode> sources = node.getSources();
+            if (sources.isEmpty()) {
+                return new Fragments(
+                        ImmutableSet.of(),
+                        ImmutableSet.of(),
+                        // fragment is considered lazy by default
+                        true);
+            }
+
+            List<Fragments> sourceFragments = node.getSources().stream()
+                    .map(subPlanNode -> subPlanNode.accept(this, currentFragmentId))
+                    .collect(toImmutableList());
+
+            return new Fragments(
+                    sourceFragments.stream()
+                            .flatMap(source -> source.getAllFragments().stream())
+                            .collect(toImmutableSet()),
+                    sourceFragments.stream()
+                            .flatMap(source -> source.getLazyFragments().stream())
+                            .collect(toImmutableSet()),
+                    sourceFragments.stream().allMatch(Fragments::isCurrentFragmentLazy));
+        }
+
+        private void addEdges(Set<PlanFragmentId> sourceFragments, Set<PlanFragmentId> targetFragments)
+        {
+            for (PlanFragmentId targetFragment : targetFragments) {
+                for (PlanFragmentId sourceFragment : sourceFragments) {
+                    fragmentDependency.addEdge(sourceFragment, targetFragment);
+                }
+            }
+        }
+    }
+
+    private static class Fragments
+    {
+        /**
+         * All upstream fragments (excluding current one)
+         */
+        private final Set<PlanFragmentId> allFragments;
```

Comment:
upstreamFragments?

Reply:
It looks like in some cases it does include self, and in some cases it doesn't. I wonder how difficult is to make it consistent (or at least fix the naming to make it less obscure)

Reply:
It is a bit confusing. It does not include `self` in return values of `visit*` methods in visitor. But it does in return of `processFragment`.